src/util/xeval/evaluator.rs

```diff
@@ -303,6 +305,31 @@ impl Evaluator {
         let (left, right) = try!(self.eval_two_children(ctx, expr));
         eval_arith(left, right, f)
     }
+
+    fn eval_case_when(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        for chunk in expr.get_children().chunks(2) {
+            let res = try!(self.eval(ctx, &chunk[0]));
+            if chunk.len() == 1 {
```

Comment:
Actually they are same. TiDB loops over the collection by step 2; here use chunks. And the loop in TiDB will end if the size is an odd number; chunks will still pass the last element in that case.